internal/scheduler/legacyscheduler.go

```diff
@@ -114,196 +177,300 @@ func queuedJobsIteratorLoader(ctx context.Context, jobIds []string, ch chan *api
+type QueuedGangIterator struct {
+	ctx                context.Context
+	queuedJobsIterator *QueuedJobsIterator
+	// Jobs are grouped into gangs by this annotation.
+	gangIdAnnotation string
+	// Jobs in a gang must specify the total number of jobs in the gang via this annotation.
+	gangCardinalityAnnotation string
+	// Groups jobs by the gang they belong to.
+	jobsByGangId map[string][]*api.Job
 }
 
-func NewQueueCandidateJobsIterator(
-	ctx context.Context,
-	queue string, initialTotalQueueResources schedulerobjects.ResourceList, initialTotalQueueResourcesByPriority schedulerobjects.QuantityByPriorityAndResourceType, scheduler LegacyScheduler,
-) (*QueueCandidateJobsIterator, error) {
-	jobsIterator, err := NewQueuedJobsIterator(ctx, queue, scheduler.JobRepository)
-	if err != nil {
-		return nil, err
+func NewQueuedGangIterator(ctx context.Context, it *QueuedJobsIterator, gangIdAnnotation, gangCardinalityAnnotation string) *QueuedGangIterator {
+	return &QueuedGangIterator{
+		ctx:                       ctx,
+		queuedJobsIterator:        it,
+		gangIdAnnotation:          gangIdAnnotation,
+		gangCardinalityAnnotation: gangCardinalityAnnotation,
+		jobsByGangId:              make(map[string][]*api.Job),
 	}
-	return &QueueCandidateJobsIterator{
-		LegacyScheduler:               scheduler,
-		ctx:                           ctx,
-		jobsIterator:                  jobsIterator,
-		totalQueueResources:           initialTotalQueueResources.DeepCopy(),
-		totalQueueResourcesByPriority: initialTotalQueueResourcesByPriority.DeepCopy(),
-		roundQueueResources:           schedulerobjects.ResourceList{},
-	}, nil
 }
 
-// Update the internal state of the iterator to reflect that a job was leased.
-func (it *QueueCandidateJobsIterator) Lease(jobSchedulingReport *JobSchedulingReport) {
-	it.totalQueueResources = jobSchedulingReport.TotalQueueResources
-	it.roundQueueResources = jobSchedulingReport.RoundQueueResources
-	it.totalQueueResourcesByPriority = jobSchedulingReport.TotalQueueResourcesByPriority
-}
+func (it *QueuedGangIterator) Next() ([]*api.Job, error) {
+	if it.jobsByGangId == nil {
+		it.jobsByGangId = make(map[string][]*api.Job)
+	}
 
-func (it *QueueCandidateJobsIterator) Next() (*JobSchedulingReport, error) {
-	// Once this function has returned error,
-	// it will return this error on every invocation.
-	if it.err != nil {
-		return nil, it.err
+	// Get one job at a time from the underlying iterator until we either
+	// 1. get a job that isn't part of a gang, in which case we yield it immediately, or
+	// 2. get the final job in a gang, in which case we yield the entire gang.
+	for {
+		job, err := it.queuedJobsIterator.Next()
+		if err != nil {
+			return nil, err
+		}
+		if job == nil {
+			return nil, nil
+		}
+
+		gangId, gangCardinality, isGangJob, err := GangIdAndCardinalityFromJob(job, it.gangIdAnnotation, it.gangCardinalityAnnotation)
+		if err != nil {
+			log := ctxlogrus.Extract(it.ctx)
```

Comment:
as discussed we should just fail the job here, but if we can't do this quickly, let's return `[]*api.Job{job}, nil` as otherwise we end up making decisions based on values returned from a function that has errored.

Reply:
this can be done as a subsequent pr